src/Confluent.Kafka/Producer.cs

```diff
@@ -17,348 +17,215 @@
-            {
-                LibRdKafka.conf_set_dr_msg_cb(configPtr, DeliveryReportCallback);
-            }
+            this.ownedClient = null;
+            this.handle = handle;
+            this.producer = (Producer)handle.Owner;
+            setAndValidateSerializers(keySerializer, valueSerializer);
+        }
 
-            // Explicitly keep references to delegates so they are not reclaimed by the GC.
-            errorDelegate = ErrorCallback;
-            logDelegate = LogCallback;
-            statsDelegate = StatsCallback;
 
-            // TODO: provide some mechanism whereby calls to the error and log callbacks are cached until
-            //       such time as event handlers have had a chance to be registered.
-            LibRdKafka.conf_set_error_cb(configPtr, errorDelegate);
-            LibRdKafka.conf_set_log_cb(configPtr, logDelegate);
-            LibRdKafka.conf_set_stats_cb(configPtr, statsDelegate);
+        /// <summary>
+        ///     Refer to <see cref="Confluent.Kafka.IClient.Name" />
+        /// </summary>
+        public string Name
+            => this.handle.Owner.Name;
 
-            this.kafkaHandle = SafeKafkaHandle.Create(RdKafkaType.Producer, configPtr);
-            configHandle.SetHandleAsInvalid(); // config object is no longer useable.
 
-            if (!manualPoll)
-            {
-                callbackCts = new CancellationTokenSource();
-                callbackTask = StartPollTask(callbackCts.Token);
-            }
+        internal int Flush(int millisecondsTimeout)
+            => ((Producer)this.handle.Owner).Flush(millisecondsTimeout);
 
-            // note: ConcurrentDictionary.GetorAdd() method is not atomic
-            this.topicHandlerFactory = (string topicName) =>
-            {
-                // Note: there is a possible (benign) race condition here - topicHandle could have already
-                // been created for the topic (and possibly added to topicHandles). If the topicHandle has
-                // already been created, rdkafka will return it and not create another. the call to rdkafka
-                // is threadsafe.
-                return kafkaHandle.Topic(topicName, IntPtr.Zero);
-            };
-        }
 
         /// <summary>
-        ///     Initializes a new Producer instance.
+        ///     Wait until all outstanding produce requests and delievery report
+        ///     callbacks are completed.
+        ///    
+        ///     [API-SUBJECT-TO-CHANGE] - the semantics and/or type of the return value
+        ///     is subject to change.
         /// </summary>
-        /// <param name="config">
-        ///     librdkafka configuration parameters (refer to https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md).
+        /// <param name="timeout">
+        ///     The maximum length of time to block. You should typically use a
+        ///     relatively short timout period because this operation cannot be
+        ///     cancelled.
```

Comment:
If suggesting to use a short timeout, then also suggest that they call this repeatedly until it reaches 0, otherwise there will be message loss